### ISSUE TITLE
Enhancement: [naming-convention] Allow typeLike imports and variableLike imports to use separate formats

### ISSUE THREAD

**Issue by benblank** [typescript-eslint/typescript-eslint#8429]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/naming-convention/

### Description

Having separate naming conventions for type-like imports (even those without TypeScript's `type` keyword, such as classes used as values) and variable-like imports would allow imported identifiers to have the same level of naming restrictions as non-imported identifiers, e.g. ensuring type-like identifiers use PascalCase.

The pass/fail examples below assume an ESLint config similar to the following (which reimagines "import" as a modifier rather than a selector only because that's the model which feels most natural to me).

``` json
{
  "rules": {
    "@typescript-eslint/naming-convention": ["error", [
      { "selector": "typeLike", "modifiers": ["import"], "format": ["PascalCase"] },
      { "selector": "variableLike", "modifiers": ["import"], "format": ["camelCase"] }
    ]]
  }
}
```

### Fail

```TypeScript
import eventEmitter from "node:events";
import { readFile as ReadFile } from "node:fs";
```

### Pass

```TypeScript
import EventEmitter from "node:events";
import { readFile } from "node:fs";
```

**Comment by bradzacher** [comment]
In general we want the rule to act in a single-file. Why? Because otherwise changing one file forces you to rename things across other files - which we can't automate and thus needs to be actioned manually by the user.

In the vast majority of usecase we've found that users don't want to rename imports because it's the exported names that inform the imported names - I.e. There's no reason to enforce a specific convention for imported names because for first party code it's already correct.

For 3rd party code similarly people prefer to not rename them because the syntax is verbose and manual. Additionally changing the imported names can easily make it hard to search the codebase because people mightn't use the same name.

In fact that has been one of the bigger things people have asked for - people generally prefer to use keep the 3rd party naming because it aids with searching docs and such.

Finally - cross file information like this requires type information. We are very careful when adding type information to a rule that doesn't use it because it restricts people from using a rule. We want very good reasons for doing so (generally because it fixes a bug or some correctness thing).

Based on the above and based on past user experience - I'm not convinced that this usecase passed the "useful to the broader TypeScript community" test.

**Comment by benblank** [comment]
Named exports indeed aren't a good use case for this; in retrospect, I'm not sure why I even mentioned them. 🙁

Cross-file information is certainly a concern, but the rule does already use type information (at least conditionally; I haven't dug deeply enough into the source to know what would happen if one were to use the `"types"` option when type information isn't available). The default options for the rule also [already enforce limiting default import names][default-imports] to specific formats.

[default-imports]: https://typescript-eslint.io/play/#code=JYWwDg9gTgLgBAUQGoIHIBUD6CCyBJddBAJTgDMoIQ4AiAOwgBMBTALmYDdm6YBnGgNwBYAFBA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6JgQwFtmBzWsgHsmAN0QtKw9FETRog6JHBgAviBVA

I think it's quite reasonable to want to be able to place the same level of restrictions on "name your imports whatever you want" as on "name your variables whatever you want". On that basis, I'd argue that there's broad appeal. I think the question, then, is whether that appeal outweighs the costs of introducing cross-file information. And as you say, perhaps not. 🙂

**Comment by bradzacher** [comment]
> what would happen if one were to use the "types" option when type information isn't available

The rule will crash out if you have a `types` selector configured without type-information (in the same way as any other type-aware lint rule).

> The default options for the rule also [already enforce limiting default import names](https://typescript-eslint.io/play/#code=JYWwDg9gTgLgBAUQGoIHIBUD6CCyBJddBAJTgDMoIQ4AiAOwgBMBTALmYDdm6YBnGgNwBYAFBA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6JgQwFtmBzWsgHsmAN0QtKw9FETRog6JHBgAviBVA) to specific formats.

The default options are more-or-less the same as the [camelcase](https://eslint.org/docs/latest/rules/camelcase) rule that it replaced. If you [turn the `camelcase` rule on you will see it report by default](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.ts&code=PTAEAEBcGcFoEsDmA7A9gJwKYCh4FsAHDSUTAN02UgH1M95JJN1QAzdVPUAIjQBNMALnKUY3ANzZsQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQMYEMBbReA5RdKRaaAe2kgBpxtIABAFwE8AHFXaAEseHALQp4ggHYcA9FKLSA5qNy0pAN0QzB6ypGp0GLAL4gTQA&tsconfig=&tokens=false). The `naming-convention` config is slightly stricter in that it doesn't allow `UPPER_CASE`, however.

The idea is that a long time ago we had a version of `camelcase` that supported TS constructs alongside a few other naming related rules and they all enforced different options for different constructs. `naming-convention` was the unification of all of the rules to have the same set of features and it replaced them all. The default config is there so people can migrate from `camelcase: 'error'` to `'@typescript-eslint/naming-convention': 'error'` without (mostly) needing to do anything as the rule is a complex one to fully configure.

> On that basis, I'd argue that there's broad appeal

When we say "broad appeal" we mean something along the lines of "a significant portion of users will want to use this option". From our anecdotal experience people generally want to just leave imports how they are defined and don't want to faff around with renaming if they can avoid it. This is evidenced by fact the rule existed for over 4 years before it even had support for checking imported names (that support only landed ~4 months ago!) - very few people asked for support.

---

Based on the above - I'm going to reject this request.
The cost of type-info far outweighs the benefit, and we don't believe that there is going to be broad adoption of the feature based on the above.
